### ISSUE TITLE
Can't navigate grouped list calendars that list group of n events

### ISSUE THREAD

**Issue by nekojira** [moonstonemedia/WP-Google-Calendar-Events#52]
Make a calendar with 'grouped list' view.

Unlike the normal 'list' view which is monthly, the grouped list can group together and arbitrary amount of months, weeks or days of events. But also group together events, regardless of time range. This is to port a feature found in 2.x where this seems currently possible.

If grouping together days, weeks or months, the calendar seem to display the right amount and navigate across pages that hold the same amount of events (maybe it'd be helpful to test when the calendar approaches the last page and the number in the chunk changes).

What doesn't work well it's when the calendar groups together events. It fetches the right amount of initial events, but then doesn't navigate through pages. Here is the PHP part when it builds the events list for the current page (which works, but only for the first page):

https://github.com/moonstonemedia/WP-Google-Calendar-Events/blob/refactor/includes/calendars/views/default-calendar-list-grouped.php#L258

And here is the javascript counterpart (which works also for the other groups):

https://github.com/moonstonemedia/WP-Google-Calendar-Events/blob/refactor/assets/js/gce-default-calendar.js#L99

it passes a timestamp to the ajax callback and then the php should do the rest

this doesn't seem a javascript problem, the response of the callback is always the same - it's likely there's a problem in how the list of events is calculated by slicing the events array?

**Comment by nekojira** [comment]
@nickyoung87  got more fun for you :tada:

**Comment by nekojira** [comment]
grouped list calendar is just a duplicate of the regular list

after the navigation issue is sorted, going to delete the regular list and keep only this grouped (renamed in just 'list') with 1 month span as default setting == current monthly list

**Comment by nekojira** [comment]
this seems to have improved with latest commits, but needs testing before closing it

**Comment by nekojira** [comment]
in both default calendar views, the navigation needs to know what's previous/next

for grid, this was relatively easier as it's a monthly calendar

but with lists, list can be monthly, bi-monthly, show chunks of events on a n-week basis or even display group of events unrelated to a date range, but in a sequence

so, for lists, there are three data fields that the js reads and they're all timestamps - current, prev and next. Current, along the direction (forward or backward?) is passed to ajax callback, and all the calculations to what to show next or previous is done by php. 

this is how it determines the next/prev block in a events sequence when a **period range** is selected for the list (days, weeks, months...):

https://github.com/moonstonemedia/WP-Google-Calendar-Events/blob/refactor/includes/calendars/views/default-calendar-list.php#L232-L243

whereas this is how it does it when instead blocks of n events are requested per page:

https://github.com/moonstonemedia/WP-Google-Calendar-Events/blob/refactor/includes/calendars/views/default-calendar-list.php#L267-L270

important properties are start, end (of the current page) as well as calendar->start and calendar->end  these are used by js to know when navigation has to stop and the user can't proceed further or before 

the list also attempts to determine what's in the current page (months, days...) and will adjust the header accordingly. This is not perfect but so far goes by good approximation.

**Comment by nekojira** [comment]
definitely needs a bit of tweaking for all possible combinations (dates, timezones, but also calendar start/end and events start/end)

in the latest update tried to create a calendar that opens 'today' with latest event tomorrow; in this case the list shouldn't advance at all if there are no events tomorrow but it still seems to do that (allows advancing one day/page) - although the list of events is correct (on the second page will return empty)

**Comment by nickyoung87** [comment]
@nekojira Are the settings in place for this already? I am not sure I understand what settings you are referring to where the n would be set.

**Comment by nekojira** [comment]
precisely here: https://github.com/moonstonemedia/WP-Google-Calendar-Events/blob/refactor/includes/calendars/default-calendar.php#L241-L285

I think it's not working too bad after the last adjustments,

perhaps you can help me making the heading more precise (sometimes the months appearing in heading don't match the events listed - it's not entirely inaccurate because those are the months queried, it only needs some adjustment or offset

https://github.com/moonstonemedia/WP-Google-Calendar-Events/blob/refactor/includes/calendars/views/default-calendar-list.php#L294-L360)

**Comment by nekojira** [comment]
it is better than before but there are still bugs:

- when displaying groups of n events, while navigating back and forth the results won't be coherent/constant but will change, I think this has to do with the array keys in the links mentioned in the earlier comment

- the first page has usually a upper boundary timestamp that goes into the next month even if the events fall within the current one, always resulting in a heading that says, for example, 'September-October 2015' while it should read only 'September 2015' - browsing the next page or previous page usually doesn't have this problem, but it represents occasionally

- the next page will always go one page too further ahead, with the last page being always empty, while navigation should stop at the page before (this maybe is connected with the two problems above?)
